cocos/renderer/gfx-metal/MTLComputeCommandEncoder.h

```diff
@@ -0,0 +1,104 @@
+ of this software and associated engine source code (the "Software"), a limited,
+ worldwide, royalty-free, non-assignable, revocable and non-exclusive license
+ to use Cocos Creator solely to develop games on your target platforms. You shall
+ not use Cocos Creator software for developing other software or tools that's
+ used for developing games. You are not granted to publish, distribute,
+ sublicense, and/or sell copies of Cocos Creator.
+
+ The software or tools in this License Agreement are licensed, not sold.
+ Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+****************************************************************************/
+
+#pragma once
+
+#include "MTLCommandEncoder.h"
+#include "MTLStd.h"
+#include "MTLUtils.h"
+#include "base/Macros.h"
+#include "math/Math.h"
+#include <Metal/MTLComputePipeline.h>
+#include <Metal/MTLComputeCommandEncoder.h>
+
+namespace cc {
+namespace gfx {
+class CCMTLComputeCommandEncoder final : public CCMTLCommandEncoder {
+public:
+    CCMTLComputeCommandEncoder()                                   = default;
+    ~CCMTLComputeCommandEncoder()                                  = default;
+    CCMTLComputeCommandEncoder(const CCMTLComputeCommandEncoder &) = delete;
+    CCMTLComputeCommandEncoder(CCMTLComputeCommandEncoder &&)      = delete;
+    CCMTLComputeCommandEncoder &operator=(const CCMTLComputeCommandEncoder &) = delete;
+    CCMTLComputeCommandEncoder &operator=(CCMTLComputeCommandEncoder &&) = delete;
+
+    void initialize(id<MTLCommandBuffer> commandBuffer) {
+        _mtlEncoder  = [[commandBuffer computeCommandEncoder] retain];
+        _initialized = true;
+    }
+
+    bool isInitialized() {
+        return _initialized;
+    }
+
+    CC_INLINE void setComputePipelineState(id<MTLComputePipelineState> pipelineState) {
+        if(_pipelineState == pipelineState)
+            return;
+        [_mtlEncoder setComputePipelineState:pipelineState];
+        _pipelineState = pipelineState;
+    }
+
+    CC_INLINE void setBuffer(const id<MTLBuffer> buffer, uint offset, uint index) {
+        [_mtlEncoder setBuffer:buffer offset:offset atIndex:index];
+    }
+
+    CC_INLINE void setTexture(const id<MTLTexture> texture, uint index) {
+        [_mtlEncoder setTexture:texture atIndex:index];
+        _resourceSize = {texture.width, texture.height, texture.depth};
+    }
+    
+    CC_INLINE void dispatch(MTLSize groupsPerGrid) {
+        // nothing to do with thread count in GLSL shader
```

Comment:
GLSL shader?

Reply:
GLSL shader request to specify the compute thread size, no such limit in Metal and have to set compute thread size explicity, so clarified here